table/compaction_merging_iterator.cc

```diff
@@ -38,6 +43,17 @@ class CompactionMergingIterator : public InternalIterator {
       pinned_heap_item_[i].level = i;
       pinned_heap_item_[i].type = HeapItem::DELETE_RANGE_START;
     }
+    if (internal_stats_) {
+      TEST_SYNC_POINT("CompactionMergingIterator::UpdateInternalStats");
+      // The size of children_ or range_tombstone_iters_ (n) should not change
+      // but to be safe, we can record the size here so we decrement by the
+      // correct amount at destruction time
+      num_sorted_runs_recorded_ = n;
+      internal_stats_->IncrNumRunningCompactionSortedRuns(
+          num_sorted_runs_recorded_);
+      assert(num_sorted_runs_recorded_ <=
+             internal_stats_->NumRunningCompactionSortedRuns());
```

Comment:
Can we do assertion like this given that there can be concurrent compactions going on? Is it possible that between IncrNumRunningCompactionSortedRuns and NumRunningCompactionSortedRuns in this thread, the other thread(s) decrement a bunch of sorted runs? Same for the assertion in the destructor.

Reply:
I think we are okay with concurrent compactions:
1. we are using `std::atomic`
2. even if other threads decrement the variable, we still should never end up with `num_sorted_runs_recorded_` > `internal_stats_->NumRunningCompactionSortedRuns()` because every decrement can only cancel an increment

Reply:
NVM on my questions - turned out we could - since the operation should be guaranteed in order within the same thread